pgx/src/array.rs

```diff
@@ -1,23 +1,104 @@
 use bitvec::prelude::*;
 use bitvec::ptr::{bitslice_from_raw_parts_mut, BitPtr, BitPtrError, Mut};
 use core::ptr::{slice_from_raw_parts_mut, NonNull};
 use core::slice;
 
-extern "C" {
-    /// # Safety
-    /// Does a field access, but doesn't deref out of bounds of ArrayType
-    fn pgx_ARR_DATA_PTR(arrayType: *mut ArrayType) -> *mut u8;
-    /// # Safety
-    /// Does a field access, but doesn't deref out of bounds of ArrayType
-    fn pgx_ARR_DIMS(arrayType: *mut ArrayType) -> *mut libc::c_int;
-    /// # Safety
-    /// Must only be used on a "valid" (Postgres-constructed) ArrayType
-    fn pgx_ARR_NELEMS(arrayType: *mut ArrayType) -> i32;
-    /// # Safety
-    /// Does a field access, but doesn't deref out of bounds of ArrayType
-    fn pgx_ARR_NULLBITMAP(arrayType: *mut ArrayType) -> *mut bits8;
+#[allow(non_snake_case)]
+#[inline(always)]
+const unsafe fn TYPEALIGN(alignval: usize, len: usize) -> usize {
+    // #define TYPEALIGN(ALIGNVAL,LEN)  \
+    // (((uintptr_t) (LEN) + ((ALIGNVAL) - 1)) & ~((uintptr_t) ((ALIGNVAL) - 1)))
+    ((len) + ((alignval) - 1)) & !((alignval) - 1)
+}
+
+#[allow(non_snake_case)]
+#[inline(always)]
+const unsafe fn MAXALIGN(len: usize) -> usize {
+    // #define MAXALIGN(LEN) TYPEALIGN(MAXIMUM_ALIGNOF, (LEN))
+    TYPEALIGN(pg_sys::MAXIMUM_ALIGNOF as _, len)
+}
+
+#[allow(non_snake_case)]
+#[inline(always)]
+unsafe fn ARR_NDIM(a: *mut pg_sys::ArrayType) -> usize {
+    // #define ARR_NDIM(a)				((a)->ndim)
+    a.as_ref().unwrap_unchecked().ndim as usize
+}
+
+#[allow(non_snake_case)]
+#[inline(always)]
+unsafe fn ARR_HASNULL(a: *mut pg_sys::ArrayType) -> bool {
+    // #define ARR_HASNULL(a)			((a)->dataoffset != 0)
+    a.as_ref().unwrap_unchecked().dataoffset != 0
+}
+
+#[allow(non_snake_case)]
+#[inline(always)]
+const unsafe fn ARR_DIMS(a: *mut pg_sys::ArrayType) -> *mut i32 {
+    // #define ARR_DIMS(a) \
+    // ((int *) (((char *) (a)) + sizeof(ArrayType)))
+
+    a.cast::<u8>().add(std::mem::size_of::<pg_sys::ArrayType>()).cast::<i32>()
+}
+
+#[allow(non_snake_case)]
+#[inline(always)]
+unsafe fn ARR_NELEMS(a: *mut pg_sys::ArrayType) -> usize {
+    pg_sys::ArrayGetNItems(a.as_ref().unwrap_unchecked().ndim, ARR_DIMS(a)) as usize
+}
+
+#[allow(non_snake_case)]
+#[inline(always)]
+unsafe fn ARR_NULLBITMAP(a: *mut pg_sys::ArrayType) -> *mut pg_sys::bits8 {
+    // #define ARR_NULLBITMAP(a) \
+    // (ARR_HASNULL(a) ? \
+    // (bits8 *) (((char *) (a)) + sizeof(ArrayType) + 2 * sizeof(int) * ARR_NDIM(a)) \
+    // : (bits8 *) NULL)
+    //
+
+    if ARR_HASNULL(a) {
+        a.cast::<u8>().add(
+            std::mem::size_of::<pg_sys::ArrayType>() + 2 * std::mem::size_of::<i32>() * ARR_NDIM(a),
+        )
+    } else {
+        std::ptr::null_mut()
+    }
+}
+
+/// The total array header size (in bytes) for an array with the specified
+/// number of dimensions and total number of items.
+#[allow(non_snake_case)]
+#[inline(always)]
+const unsafe fn ARR_OVERHEAD_NONNULLS(ndims: usize) -> usize {
```

Comment:
Not unsafe, also gained an extra character here while porting.
```suggestion
const fn ARR_OVERHEAD_NONULLS(ndims: usize) -> usize {
```